sycl/source/detail/platform_impl.cpp

```diff
@@ -233,14 +235,108 @@ std::shared_ptr<device_impl> platform_impl::getOrMakeDeviceImpl(
+  return It != std::end(supported);
+}
+
+static std::vector<device> amendDeviceAndSubDevices(
+    backend PlatformBackend, std::vector<device> &DeviceList,
+    ods_target_list *OdsTargetList, int PlatformDeviceIndex) {
+  constexpr info::partition_property partitionProperty =
+      info::partition_property::partition_by_affinity_domain;
+  constexpr info::partition_affinity_domain affinityDomain =
+      info::partition_affinity_domain::next_partitionable;
+
+  std::vector<device> FinalResult;
+
+  for (unsigned i = 0; i < DeviceList.size(); i++) {
+    // device has already been screened. The question is whether it should be a
+    // top level device and/or is expected to add its sub-devices to the list.
+    device &dev = DeviceList[i];
+    bool deviceAdded = false;
+    for (ods_target target : OdsTargetList->get()) {
+      backend TargetBackend =
+          target.Backend ? target.Backend.value() : backend::all;
+      if (PlatformBackend == TargetBackend || TargetBackend == backend::all) {
+        bool deviceMatch = target.HasDeviceWildCard; // opencl:*
+        if (target.DeviceType) {                     // opencl:gpu
+          deviceMatch = ((target.DeviceType == info::device_type::all) ||
+                         (dev.get_info<info::device::device_type>() ==
+                          target.DeviceType));
+
+        } else if (target.DeviceNum) { // opencl:0
+          deviceMatch =
+              (target.DeviceNum.value() == PlatformDeviceIndex + (int)i);
+        }
+
+        if (deviceMatch) {
+          // Top level matches. Do we add it, or subdevices?
+          if (target.SubDeviceNum || target.HasSubDeviceWildCard) {
+            if (supportsPartitionProperty(dev, partitionProperty) &&
+                supportsAffinityDomain(dev, partitionProperty,
+                                       affinityDomain)) {
+              auto subDevices = dev.create_sub_devices<
+                  info::partition_property::partition_by_affinity_domain>(
+                  affinityDomain);
+              if (target.HasSubDeviceWildCard) {
+                FinalResult.insert(FinalResult.end(), subDevices.begin(),
+                                   subDevices.end());
+              } else {
+                if (subDevices.size() > target.SubDeviceNum.value()) {
+                  FinalResult.push_back(subDevices[target.SubDeviceNum.value()]);
+                } else {
+                  std::cout << "subdevice index out of bounds: " << target
+                            << std::endl;
+                }
+              }
+            } else if (target.DeviceNum ||
+                       (target.DeviceType && (target.DeviceType.value() !=
+                                              info::device_type::all))) {
+              // this device was specifically requested and yet is not
+              // partitionable.
+              std::cout << "device is not partitionable: " << target
+                        << std::endl;
```

Comment:
Should this be a `sycl::feature_not_supported` exception instead?

Reply:
No, it was decided that only parsing problems should result in thrown errors. Simply being unable to match should not throw, it should only output a warning.  

 In particular, if the use has a long set of selectors, they likely don't want any one of them to throw an error.

Reply:
Fair point. Any good way to give similar warnings when the device index is out of range?

Reply:
Out of range device number or sub-device number are supposed give a warning according to the semantics we agreed upon.

Reply:
Device-index-out-of-range emits a warning.